Day 2: “Using the Ancient Evils for Debugging” / Manuel Strehl

Boldewyn opened the pull request:
This is the pull request for the “Using the Ancient Evils for Debugging“ article.

I’m happy to incorporate any feedback, if you find things you’d like to have changed!

Boldewyn committed:
add 2025-12-02 adventcalendar article

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+That is to be taken literally. Really everything, including any closing `</plaintext>` or `</html>`, will be printed as if a rogue, unclosed `<pre>` would suddenly go haywire and slurp up the rest of the page. By the way, this makes `<plaintext>` the only non-empty element that has no end tag at all.
+
+On first sight that sounds like a really stupid superpower. On second sight, it still does. We look into how that element became part of HTML below. But now we will use it for one specific purpose: Debugging server-side code.
```
> On second sight, it still does

lol

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+## What Do We Use this Power For?
+
+That is to be taken literally. Really everything, including any closing `</plaintext>` or `</html>`, will be printed as if a rogue, unclosed `<pre>` would suddenly go haywire and slurp up the rest of the page. By the way, this makes `<plaintext>` the only non-empty element that has no end tag at all.
```
Just a thought, it felt to me like the `By the way...` thought ending the quick warning section felt better in terms of flow, and the usage section can then jump immediately into the use case of the article.


```suggestion
That is to be taken literally. Really everything, including any closing `</plaintext>` or `</html>`, will be printed as if a rogue, unclosed `<pre>` would suddenly go haywire and slurp up the rest of the page. By the way, this makes `<plaintext>` the only non-empty element that has no end tag at all.

## What Do We Use this Power For?
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+Of course, specialized debuggers like XDebug for PHP or built-in error pages like in Django take over the heavy lifting here. And even the good ol’ `print "<script>console.log('here!')</script>"` is often helpful. Those tools should be high up in your utility belt.
+
+But imagine this: You are deep down in your code chasing an elusive bug that only affects some part of the HTML output, and you want to see at a quick glance on the rendered page, where this problem appears. The fastest way is to put a quick `<plaintext>` close to the offending place, reload the page, and presto! Just scan down to where the markup starts to show through.
```
Just reads a bit clunky and long, simplification suggestion:

```suggestion
But imagine this: you’re deep in your code, chasing an elusive bug that affects only part of the HTML output, and you want to spot on the rendered page exactly where it shows up.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+## The History behind this Evil
+
+How ended this seemingly fringe feature up in all mainstream browsers? It was indeed there from the very beginning of HTML as this [historic W3C document](https://www.w3.org/History/19921103-hypertext/hypertext/WWW/MarkUp/Tags.html) of 1992 proves:
```
Loving the folkloric literary touches 😆

elisehein replied:
Loving the folkloric literary touches 😆 

```suggestion
How ended this seemingly fringe feature up in all mainstream browsers? It was indeed there from the very beginning of HTML as this [historic W3C document](https://www.w3.org/History/19921103-hypertext/hypertext/WWW/MarkUp/Tags.html) from 1992 proves:
```

Just in case the backwards word order wasn't deliberate, `how did this ... end up in` would be the correct way around, but I think it works here and adds to the ancient horror style.

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+This also tells us the reason for its invention. Back at the time Sir Tim Berners-Lee’s high-end NeXT PC that he used to write the first web browser had a quarter of the power of a hand-me-down 2009 smartphone. It was important to optimize wherever you could.
+
+Given that the early WWW was meant as a place to share scientific information, having a large blob of plain text as part of your fancy new HTML page was relatively common. The possibility to end the costly HTML parser and fall back to simply printing the remainder of the file as plain text was a powerful tool.
```
I know this is a hard ask, but it would be so cool to see an example screenshot here

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+## How Safe Are We?
+
+But the element still _is_ available in all browsers. So we need to keep at least a passing knowledge of it at the back of our minds.
```
`keep ... passing knowledge` threw me a bit, might be a bit off idiomatically?

```suggestion
But the element _is_ still available in all browsers, so it's worth keeping a bit of a working knowledge of it at the back of our minds. 
```

Not sure...

elisehein replied:
Also, if you want to lean into the folklore thread, maybe `But the element still lurks in all browsers` 🧙

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+But the element still _is_ available in all browsers. So we need to keep at least a passing knowledge of it at the back of our minds.
+
+To give you an example, how this feature could be mis-used, assume a comment function on a blog, where the commenter was able to smuggle in the string `<plaintext>`. As good developers we know never to trust users’ input, so we put the comment through a sanitizer. Let’s take a look at where things can go south from here.
```
Just a suggestion to sound more conversational/natural

```suggestion
To give you an example of how this feature could be misused, say a blog’s comment function lets someone smuggle in the string `<plaintext>`. As good developers we know never to trust users’ input, so we put the comment through a sanitizer. Let’s take a look at where things can go south from here.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+### There Goes the Sanity!
+
+We run each sanitizer in its most minimal configuration that produces any output. This is by design: Sanitizers are security products. They should produce safe output by default.
```
```suggestion
We run each sanitizer in the minimal configuration that still produces output. This is by design: sanitizers are security products. They should produce safe output by default.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
````diff
@@ -1,35 +1,406 @@
+console.log(Document.parseHTML(TEST_STRING).body.innerHTML);
+```
+
+The approach of this API is to get the nesting correct again somehow according to the HTML5 parser spec. That is, close the `<p>` and `<b>` tags, then re-open the `<b>` tag as the spec suggests. The API does not deal with the special semantics of `<plaintext>` at all, though.
````
```suggestion
The approach of this API is to somehow get the nesting conformant to the HTML5 parser spec. That is, close the `<p>` and `<b>` tags, then re-open the `<b>` tag as the spec suggests. The API does not deal with the special semantics of `<plaintext>` at all, though.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+**Result:**
+
+The classic JS sanitizer chooses to remove the `<plaintext>` and all its “content”. (I put content in quotes, because _technically_ everything after the start tag would’ve been the `<plaintext>`’s content.) DOMPurify sees to it, that the elements are properly closed. The result is identical to the one of the Sanitizer API.
```
```suggestion
The classic JS sanitizer chooses to remove the `<plaintext>` and all its “content”. (I put content in quotes, because _technically_ everything after the start tag would’ve been the `<plaintext>`’s content.) DOMPurify sees to it that the elements are properly closed. The result is identical to that of the Sanitizer API.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+**Result:**
+
+Symfony’s sanitizer has a fascinating way of moving tags around. Interesting, but at least we’ve got all elements properly closed, including the un-closeable `plaintext`.
```
```suggestion
Symfony’s sanitizer has a fascinating way of moving tags around. Interesting – but at least we’ve got all elements properly closed, including the unclosable `plaintext`.
```

Boldewyn replied:
I’ve kept the comma here. The “Interesting” isn’t meant to be emphasized more than a ”hm, ok” would be, because the actual punch line of the sentence is the `</plaintext>` closing tag.

elisehein commented on hell/adventcalendar/2025/2/index.md:
````diff
@@ -1,35 +1,406 @@
+print(nh3.clean(TEST_STRING))
+```
+
+This Rust-based sanitizer advertises its speed and conformance with the HTML spec.
+
+**Result:**
````
```suggestion
**Result:**

This Rust-based sanitizer advertises its speed and conformance with the HTML spec.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+Just to be crystal clear here: we are not criticizing the result of any of these libraries. Each one has a good reason to do what they do. And each one serves a slightly different purpose.
+
+It emphasizes the point, though, that one should be absolutely sure about the aim of a chosen sanitizer and extent as to which it will change its input. Is the library for removing only potentially dangerous things but keeping as much HTML intact as possible? Is it to scrape all HTML off the string, or only escaping any HTML-special characters? The results will differ tremendously.
```
```suggestion
But it does emphasize the point that one should be absolutely sure about the aim of a chosen sanitizer and the extent to which it will change its input. Is the library for removing only potentially dangerous things but keeping as much HTML intact as possible? Is it to scrape all HTML off the string, or only escape special characters? The results will differ tremendously.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+In the case of `<plaintext>` itself we are most likely in a safe place, though. Since `<plaintext>` has built-in HTML escaping, doing something dangerous with it is severely limited. It takes quite rare a constellation of errors and oversights to appear together in order to run malicious code.
+
+Well, for the sake of the argument, let’s create such a constellation. Assume that you embed a Content-Security Policy on your site [in a `<meta>` element](https://w3c.github.io/webappsec-csp/#meta-element) instead of an HTTP header:
```
```suggestion
Just for the sake of the argument, let’s create such a constellation. Assume that you embed a Content-Security Policy on your site [in a `<meta>` element](https://w3c.github.io/webappsec-csp/#meta-element) instead of an HTTP header:
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
````diff
@@ -1,35 +1,406 @@
+<meta http-equiv="Content-Security-Policy" content="script-src 'self'">
+```
+
+This prevents loading 3rd party scripts sufficiently. If an attacker finds a possibility to load HTML prior to this element, they can nullify the CSP:
````
```suggestion
This sufficiently prevents loading 3rd party scripts. If an attacker finds a way to load HTML prior to this element, they can nullify the CSP:
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+That is to be taken literally. Really everything, including any closing `</plaintext>` or `</html>`, will be printed as if a rogue, unclosed `<pre>` would suddenly go haywire and slurp up the rest of the page. By the way, this makes `<plaintext>` the only non-empty element that has no end tag at all.
+
+On first sight that sounds like a really stupid superpower. On second sight, it still does. We look into how that element became part of HTML below. But now we will use it for one specific purpose: Debugging server-side code.
```
> On second sight, it still does.

lol 😆


```suggestion
At first glance, that sounds like a really stupid superpower. At second glance, it still does. We look into how that element became part of HTML below. But first, let's put it to work for one specific purpose: debugging server-side code.
```

Boldewyn replied:
I’ve kept the “sight”, but what I _actually_ wanted to use is “thought”. That sounds, however, even stranger.

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+On first sight that sounds like a really stupid superpower. On second sight, it still does. We look into how that element became part of HTML below. But now we will use it for one specific purpose: Debugging server-side code.
+
+Of course, specialized debuggers like XDebug for PHP or built-in error pages like in Django take over the heavy lifting here. And even the good ol’ `print "<script>console.log('here!')</script>"` is often helpful. Those tools should be high up in your utility belt.
```
```suggestion
Of course, specialized debuggers like XDebug for PHP or built-in error pages built-in error pages in frameworks like Django take over the heavy lifting here. And even the good ol’ `print "<script>console.log('here!')</script>"` is often helpful. Those tools should be high up in your utility belt.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+But imagine this: You are deep down in your code chasing an elusive bug that only affects some part of the HTML output, and you want to see at a quick glance on the rendered page, where this problem appears. The fastest way is to put a quick `<plaintext>` close to the offending place, reload the page, and presto! Just scan down to where the markup starts to show through.
+
+This is especially useful to access formatted debugging output. A `var_dump()` in PHP, for example. Or an `error.stack` stack trace in NodeJS. Slap a `<plaintext>` in front of it before writing it to the HTML output, so that the string is immediatelly readable:
```
```suggestion
This is especially useful to access formatted debugging output. A `var_dump()` in PHP, for example. Or an `error.stack` stack trace in NodeJS. Slap a `<plaintext>` in front of it before sending it to the browser, making the string is immediately readable:
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+>
+> This tag allows the rest of a file to be read efficiently without parsing. Its presence is an optimisation. There is no closing tag.
+
+This also tells us the reason for its invention. Back at the time Sir Tim Berners-Lee’s high-end NeXT PC that he used to write the first web browser had a quarter of the power of a hand-me-down 2009 smartphone. It was important to optimize wherever you could.
```
```suggestion
This gives us a clue as to why it was introduced in the first place. Back at the time, Sir Tim Berners-Lee’s high-end NeXT PC that he used to write the first web browser had a quarter of the power of a hand-me-down 2009 smartphone. It was important to optimize wherever you could.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+Given that the early WWW was meant as a place to share scientific information, having a large blob of plain text as part of your fancy new HTML page was relatively common. The possibility to end the costly HTML parser and fall back to simply printing the remainder of the file as plain text was a powerful tool.
+
+Such a feature isn’t unheard of in the world of computers, too. For example, the programming language Perl uses a [special marker](https://perldoc.perl.org/perldata#Special-Literals) to tell the Perl parser to stop processing the remainder of the file:
```
`world of computers` threw me a bit, I know what you mean here, but isn't HTML also `world of computers`? 

```suggestion
Such a feature isn’t unique to HTML, either. For example, the programming language Perl uses a [special marker](https://perldoc.perl.org/perldata#Special-Literals) to tell the Perl parser to stop processing the remainder of the file:
```

Boldewyn replied:
Good point! Thank you!

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+We run each sanitizer in its most minimal configuration that produces any output. This is by design: Sanitizers are security products. They should produce safe output by default.
+
+The results are in. And they are quite surprising. (Click on a name to read the horrifying details.)
```
The "horrifying details" setup made me expect that each sanitizer output would be individually alarming, but you don't really go into explaining what makes each of them bad. If I'm getting it right, further down it becomes clear that it's actually not that any of them are bad in any specific way, but the "horror" of it is that they're so inconsistent?

Maybe a clarification for that upfront would be good? As a reader, seeing "read these horrifying details", and then having the first example make me go "hmm, what was so bad about this?" made me feel a little stupid.

Maybe a small change along these lines would be good:

```suggestion
The results are in. And the inconsistencies are quite surprising. (Click on a name to read the horrifying details.)
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+**Result:**
+
+However, the result is close but still different to what browsers will do. In this case, it’s the `<b>` tag that would not extend over the content of the `<plaintext>` element.
```
`still` already conveys the contrasting thought.

```suggestion
The result is close but still different to what browsers will do. In this case, it’s the `<b>` tag that would not extend over the content of the `<plaintext>` element.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+---
+
+With **11** sanitizing methods we managed to produce **10** different outputs.
```
Just stylistic choice, take it or leave it!

```suggestion
With **11** sanitizing methods we managed to produce **10** different outputs!
```

Just think this is a really nice overview and deservers a little pause!

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+We enter the danger zone when mixing several tools together without taking a cautious look first. For example, look at how DOMPurify and HTML Purifier would interact in a potentially hazardous way. DOMPurify would remove any `<plaintext>` including its content. A later check for any malicious payload would be negative.
+
+HTML Purifier on the other hand just strips the `<plaintext>` tag, while its content remains on the page. If we’d trust the previous DOMPurify result, we’d be surprised by sudden new content being placed verbatim in the HTML code.
```
Slightly better paragraph break placement for flow, maybe?

```suggestion
We enter the danger zone when mixing several tools together without taking a cautious look first.

For example, look at how DOMPurify and HTML Purifier would interact in a potentially hazardous way. DOMPurify would remove any `<plaintext>` including its content. A later check for any malicious payload would be negative. HTML Purifier on the other hand just strips the `<plaintext>` tag, while its content remains on the page. If we’d trust the previous DOMPurify result, we’d be surprised by sudden new content being placed verbatim in the HTML code.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+HTML Purifier on the other hand just strips the `<plaintext>` tag, while its content remains on the page. If we’d trust the previous DOMPurify result, we’d be surprised by sudden new content being placed verbatim in the HTML code.
+
+If one library is used for input validation and another one for output quoting, this is a [cross-site scripting](https://en.wikipedia.org/wiki/Cross-site_scripting) desaster waiting to happen, unless we know _exactly_ what we’re doing.
```
```suggestion
If one library is used for input validation and another one for output quoting, that's a [cross-site scripting](https://en.wikipedia.org/wiki/Cross-site_scripting) disaster waiting to happen, unless we know _exactly_ what we’re doing.
```

Boldewyn replied:
Oh, interesting! It’s “Desaster” in German, but “disaster” in English. Hadn’t noticed that consciously before.

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+
+## Letting the Evil Sleep Again
+
+In the case of `<plaintext>` itself we are most likely in a safe place, though. Since `<plaintext>` has built-in HTML escaping, doing something dangerous with it is severely limited. It takes quite rare a constellation of errors and oversights to appear together in order to run malicious code.
```
```suggestion
In the case of `<plaintext>` itself we are most likely in a safe place. Since `<plaintext>` has built-in HTML escaping, doing something dangerous with it is severely limited. It takes quite rare a constellation of errors and oversights to appear together in order to run malicious code.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+- The CSP is not set via HTTP
+- The complete remainder of the page is converted to `text/plain`, which makes this definitively not a stealthy attack
+
+So we can conclude: It is important to know about `<plaintext>`. But if we follow tried and tested security rules (for example the [OWASP Application Security Verification Standard](https://owasp.org/www-project-application-security-verification-standard/)), we will remain safe from this ancient evil.
```
```suggestion
So, to sum up – it's important to know about `<plaintext>`. But if we follow tried and tested security rules (for example the [OWASP Application Security Verification Standard](https://owasp.org/www-project-application-security-verification-standard/)), we will remain safe from this ancient evil.
```

elisehein replied:
I do love that you mention "ancient evil" again to tie it all together. Maybe it'd be nice to go a bit further and add a nod to the fact that it is "ancient"?

I'd also add some detail to summarise the big point from the article – that sanitizers don't agree on how to handle `<plaintext>`.

```suggestion
To sum up – every era of the web leaves behind strange relics like `<plaintext>`, and it's important to know about the ghosts that haunt the platform. This one is especially mischievous: sanitizers don't agree on how to treat it, so it pays to know its quirks. Follow sound security practices (like those in [OWASP ASVS](https://owasp.org/www-project-application-security-verification-standard/)), and we'll remain safe from this ancient evil.
```

Might be going a tad too far into the metaphor, up to you 😜

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
+The results are in. And they are quite surprising. (Click on a name to read the horrifying details.)
+
+<details>
+<summary>The new HTML Sanitizer API as implemented in Firefox</summary>
```
Just a formatting thought – maybe bold all of the `<summary>` sanitiser titles so there's a bit more hierarchy when they're open?

elisehein commented on hell/adventcalendar/2025/2/index.md:
```diff
@@ -1,35 +1,406 @@
-Some text.
 
-Some text. Some text.
+Deep down in the dark voids of HTML specs long gone sleeps a terrifying thing. Imagine, if you will, a DOM node so mighty, that it can change the `content-type` of parts of the document. An HTML element that makes the parser tremble and withdraw, and that cannot be stopped even by its own end tag.
```
```suggestion
Deep down in the dark voids of HTML specs long gone sleeps a terrifying thing. Imagine, if you will, a DOM node so mighty, that it can change the document’s `content-type` mid-stream. An HTML element that makes the parser tremble and withdraw, and that cannot be stopped even by its own end tag.
```

elisehein commented on hell/adventcalendar/2025/2/index.md:
````diff
@@ -1,35 +1,406 @@
+```
+
+![A screenshot of the HTMHell website where the lower part shows a PHP variable output followed by the site’s markup instead of the rendered HTML](./debug_php.png)
+
````
Great example, but a little brief – maybe a one-liner showing the same trick in JavaScript or Node would help generalize it?

elisehein reviewed (commented):
Hi Manuel, I'm Elise 👋 I'm helping the team review some of the advent calendar articles. 

I didn't know about `<plaintext>` and learned a lot! The "ancient evil" thread running throughout the post was solid without turning cringe, and it was structurally really easy to follow. The examples helped. Great job making a long technical deep dive feel like a story!

Most of my comments are nits around more idiomatic / natural sounding English, a few restructuring suggestions and a couple of typos. Ignore at will!

A few comments on the content:

* Already mentioned below, but it wasn't entirely clear when we're diving into the different sanitisers what I should keep an eye out for, we just go through them without any critique – it only becomes clear later that the fact they all differ is the surprising bit. Further, that in itself assumes readers understand why inconsistent sanitizer output matters. I don't want to suggest you change your narrative if it was intentional that the "big reveal" comes later, but it would have been helpful for me to have more context upfront before going into the examples. Something like `Sanitizers are meant to guarantee safety, but as you’ll see, they don’t even agree on what safety means. That inconsistency can lead to subtle XSS risks when tools are combined.`
* Maybe re-word the parts of the piece that mention content-type – from what I'm learning, technically `plaintext` doesn’t change MIME type, it just switches the parser’s insertion mode. It's not critical to the argument, but could clarify for nuance.
* You do show a screenshot of what the output of `plaintext` looks like, but maybe for the non-experts an even more basic before/after illustration would be useful early on.

Boldewyn commented:
Hi Elise!
Thank you very much for the thorough review and both the nice words and the great critique!

The elephant in the room is the part directly before and after listing the sanitizer results. I wasn’t too happy with it, but when the deadline approached I just rolled with it. Good to know, that that’s indeed a part that needs more work! I’ll take a look!

(This part with all the sanitizers wasn’t actually planned from the get-go. I was just “huh? That’s strange!” kind of surprised, when my choice of two sanitizers returned very different results and then started to look at all the others, noticing the diversity in responses. It’s therefore more or less original research, but it still feels a bit tacked-on.)

Re: historic screenshot: OK, that one was hard, but I remembered https://worldwideweb.cern.ch/code/, a project where they tried to recreate the very first web browser, and https://www.w3.org/History/19921103-hypertext/hypertext/WWW/MarkUp/Connolly/structure5.html, which is an actual example from 1992 showing the use of `<plaintext>`. I managed to get the first to render the latter. I’ll add the screenshot to the article.

There’s one thing that Tom Schuster (one of the Sanitizer API spec authors) mentioned to me, too: The “poor man’s sanitizing” is no sanitizing at all. I’ll include a bit of clarification in that part as well.

matuzo commented:
Elise already gave you lots of feedback; I have nothing more to add. Interesting topic and easy to read. Great job!

Boldewyn committed:
incorporate feedback to 2025-02 adventcalendar article

Many thanks to @elisehein for the thorough review!

Boldewyn commented:
I’ve updated the post based on Elise’s feedback.

@elisehein I added you to the ”thank you” section at the end. Is that OK with you, and if so, do you want a different URL linked with your name?

elisehein commented:
Thanks @Boldewyn, that's very kind. Feel free to link to https://elisehe.in.

Boldewyn committed:
Update link in 2025-12-02 adventcalendar article

matuzo commented:
@Boldewyn Is the post ready, can I merge it?

Boldewyn commented:
Yes, ready to be merged. :+1:
